File: SearchInSorted2DMatrix.java
/**
Problem: Search in a 2-D sorted matrix.
https://leetcode.com/problems/search-a-2d-matrix/

Time Complexity : log (m * n)
Space Complexity : O(1)
Did this code successfully run on Leetcode : yes
Any problem you faced while coding this : No
*/

class SearchInSorted2DMatrix {
    public boolean searchMatrix(int[][] matrix, int target) {
        int rows = matrix.length;
        
        if (rows == 0) {
            return false;
        }
        
        int cols = matrix[0].length;
        if (cols == 0) {
            return false;
        }
        
        // Treat the 2d matrix as a 1d sorted array and perform binary search on it.
        int low = 0, high = rows * cols - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            // Get convert index of mid in 1d array to 2d array
            if (matrix[mid / cols][mid % cols] == target) {
                return true;
            } else if (target > matrix[mid / cols][mid % cols]) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return false;
    }
}

File: search-in-sorted-2d-matrix.py
# time complexity: O(log(m)+log(n)) or O(log(m*n)) where m = # of rows and n = # of columns in a row
# space complexity: O(1) as no additional space used
# # Did this code successfully run on Leetcode : yes
# Any problem you faced while coding this : confused on the formula we have used to find the vector for mid element
# i.e. mid_row = mid//n and mid_column = mid%n

class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """

        # number of rows
        m = len(matrix)

        # number of columns
        n = len(matrix[0])

        # consider the 2D matrix as 1D list of list of int with total length m*n
        lo = 0
        hi = (m*n)-1

        while lo <= hi:
            mid = (lo + hi) // 2

            # find vector of mid idx
            mid_row = mid//n
            mid_col = mid % n

            # normal BS
            if (matrix[mid_row][mid_col] == target):
                return True
            elif (matrix[mid_row][mid_col] > target):
                hi = mid - 1
            else:
                lo = mid + 1

        return False    # return false if element not found
